### DISCUSSION TITLE
Upgrade to Jakarta EE 10 and Java 17

### DISCUSSION THREAD

**Discussion by hantsy** [micronaut-projects/micronaut-core#8557] [General]
### Feature description

Currently most of the Microant modules are based on JavaEE 8/Jakarta EE 8, Jakarta EE 10 was released for a long time, and the latest Spring 6/Spring Boot 3.0 has already upgraded to Jakarta EE9/10 and Java 17. 

I think in the next generation 4.0 Micronaut should consider the alignment of Jakarta EE 10 and Java 17.

Upgrade all modules to use new `jakarta.`. 

* Jakarta Common Annotations 
* Jakarta Bean Validation 3.0
* Jakarta Persistence 3.1.0
* Implements CDI Lite 4.0
* Jakarta Transaction 2.0
* JSON-B/JSON-P as Jackson alternatives

Use Java 17 as base line to build Micronaut projects, consider more integration with new features in Java 17, such as Record, HttpClient(since Java 11) etc.

**Comment by sdelamo** [comment]
@hantsy Thanks for the feedback. Current plan is to have Micronaut Framework 4 with Java 17 as a baseline. 

We have been upgrading as much as possible to use `jakarta.` as well. 

moving this to discussions.

**Comment by sdelamo** [comment]
@hantsy we discussed the baseline here: https://github.com/micronaut-projects/micronaut-core/issues/7006

**Reply by dstepanov** [reply]
It's the plan.

>  Jakarta Common Annotations

You can use them in Micronaut 3

>  Jakarta Bean Validation 3.0

There is work in progress targeted for Micronaut 4

>  Jakarta Persistence 3.1.0
> Jakarta Transaction 2.0

You can use them in the latest Micronaut Data 3 for Micronaut JDBC/R2DBC. The only remaining thing is the support for Hibernate 6

>  Implements CDI Lite 4.0

There is work in progress outside of Micronaut. It will be released somehow after Micronaut 4

> JSON-B/JSON-P as Jackson alternatives

Already supported by Micronaut serialization

> Use Java 17 as base line to build Micronaut projects, consider more integration with new features in Java 17, such as Record, HttpClient(since Java 11) etc.

Java records have been supported since the preview. There is a WIP prototype of the HTTP client using Java's HttpClient internally instead of Netty.

**Reply by hantsy** [reply]
@dstepanov Thanks for explaining these in details.
